File: 백준/Gold/1987. 알파벳/알파벳.py
from sys import stdin
input = stdin.readline

d = ((-1,0),(1,0),(0,-1),(0,1))

r,c = map(int,input().split())
board = [list(input().rstrip()) for _ in range(r)]

def dfs(x,y,cnt):
    global ans
    ans = max(ans,cnt)
    for i in range(4):
        nx = x + d[i][0]
        ny = y + d[i][1]
        if 0 <= nx < r and 0 <= ny < c and board[nx][ny] not in record:
            record.add(board[nx][ny])
            dfs(nx,ny,cnt+1)
            record.remove(board[nx][ny])

ans = 0
record = set(board[0][0])
dfs(0,0,1)
print(ans)

File: 백준/Gold IV/1987. 알파벳/알파벳.java
import java.io.*;
import java.util.*;

public class Main {
	static int N, M, ans;
	static int[][] board;
	static boolean vis[] = new boolean[26];
	static final int[] dx = { -1, 0, 1, 0 };
	static final int[] dy = { 0, 1, 0, -1 };

	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		board = new int[N][M];
		for (int i = 0; i < N; i++) {
			String row = br.readLine();
			for (int j = 0; j < M; j++) {
				board[i][j] = (int) (row.charAt(j) - 'A');
			}
		}
		ans = 1;
		dfs(0, 0, 1);
		System.out.println(ans);
	}

	static void dfs(int x, int y, int cnt) {
		ans = Math.max(ans, cnt);
		vis[board[x][y]] = true;
		for (int d = 0; d < 4; d++) {
			int nx = x + dx[d];
			int ny = y + dy[d];
			if (nx >= 0 && nx < N && ny >= 0 && ny < M && !vis[board[nx][ny]]) {
				dfs(nx, ny, cnt+1);
			}
		}
		vis[board[x][y]] = false;
	}
}